tsfc/loopy.py

```diff
@@ -0,0 +1,340 @@
+            if isinstance(index, gem.Index):
+                rank.append(self.active_indices[index])
+            elif isinstance(index, gem.VariableIndex):
+                rank.append(expression(index.expression, self))
+            else:
+                assert isinstance(index, int)
+                rank.append(index)
+        return tuple(rank)
+
+    def pymbolic_variable(self, node):
+        try:
+            pym = self.gem_to_pymbolic[node]
+        except KeyError:
+            name = self.name_gen(node.name)
+            pym = p.Variable(name)
+            self.gem_to_pymbolic[node] = pym
+        if node in self.indices:
+            rank = self.pym_multiindex(self.indices[node])
+            if rank:
+                return p.Subscript(pym, rank)
+            else:
+                return pym
+        else:
+            return pym
+
+    def active_inames(self):
+        # Return all active indices
+        return frozenset([i.name for i in self.active_indices.values()])
+
+
+def generate(impero_c, args, precision, kernel_name="loopy_kernel", index_names=[]):
+    """Generates loopy code.
+
+    :arg impero_c: ImperoC tuple with Impero AST and other data
+    :arg args: list of loopy.GlobalArgs
+    :arg precision: floating-point precision for printing
+    :arg kernel_name: function name of the kernel
+    :arg index_names: pre-assigned index names
+    :returns: loopy kernel
+    """
+    ctx = LoopyContext()
+    ctx.indices = impero_c.indices
+    ctx.index_names = defaultdict(lambda: "i", index_names)
+    ctx.precision = precision
+    ctx.epsilon = 10.0 ** (-precision)
+
+    # Create arguments
+    data = list(args)
+    for i, temp in enumerate(impero_c.temporaries):
+        name = "t%d" % i
+        if isinstance(temp, gem.Constant):
+            data.append(lp.TemporaryVariable(name, shape=temp.shape, dtype=temp.array.dtype, initializer=temp.array, scope=lp.temp_var_scope.GLOBAL, read_only=True))
+        else:
+            shape = tuple([i.extent for i in ctx.indices[temp]]) + temp.shape
+            data.append(lp.TemporaryVariable(name, shape=shape, dtype=numpy.float64, initializer=None, scope=lp.temp_var_scope.LOCAL, read_only=False))
+        ctx.gem_to_pymbolic[temp] = p.Variable(name)
+
+    # Create instructions
+    instructions = statement(impero_c.tree, ctx)
+
+    # Create domains
+    domains = []
+    for idx, extent in ctx.index_extent.items():
+        inames = isl.make_zero_and_vars([idx])
+        domains.append(((inames[0].le_set(inames[idx])) & (inames[idx].lt_set(inames[0] + extent))))
+
+    if not domains:
+        domains = [isl.BasicSet("[] -> {[]}")]
+
+    # Create loopy kernel
+    knl = lp.make_kernel(domains, instructions, data, name=kernel_name, target=lp.CTarget(), seq_dependencies=True)
+
+    # Prevent loopy interchange by loopy
+    knl = lp.prioritize_loops(knl, ",".join(ctx.index_extent.keys()))
+
+    # Help loopy in scheduling by assigning priority to instructions
```

Comment:
Why not prioritise before making the kernel?

Reply:
Mainly that I don't know how many instructions I have in total, and it's nicer to have priorities go down from `len(instructions)` down to 0?

Reply:
OK.

Reply:
Probably need to change this after we do dependencies correctly. Loopy doesn't seem to guarantee deterministic code generation though, which will be a problem.